features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c

```diff
@@ -0,0 +1,333 @@
+         */
+        if( cmac_ctx->unprocessed_len == SASI_AES_BLOCK_SIZE_IN_BYTES && ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_block,
+                              SASI_AES_BLOCK_SIZE_IN_BYTES, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+            cmac_ctx->unprocessed_len = 0;
+        }
+    }
+
+    if( ilen > 0 )
+    {
+        const size_t size_to_store = ( ilen % SASI_AES_BLOCK_SIZE_IN_BYTES == 0) ?
+                                     SASI_AES_BLOCK_SIZE_IN_BYTES : ilen % SASI_AES_BLOCK_SIZE_IN_BYTES;
+        memcpy( &cmac_ctx->unprocessed_block[0],
+                           input + ilen - size_to_store,
+                           size_to_store );
+        cmac_ctx->unprocessed_len = size_to_store;
+        ilen -= size_to_store;
+        if( ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, ( uint8_t * )input,
+                               ilen, NULL ) != 0 )
+            {
+                ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+                goto exit;
+            }
+        }
+    }
+
+exit:
+    if( ret != 0 )
+    {
+        mbedtls_platform_zeroize( cmac_ctx, sizeof( *cmac_ctx ) );
+        mbedtls_free( cmac_ctx );
+        SaSi_AesFree( &cmac_ctx->CC_Context );
+    }
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_finish( mbedtls_cipher_context_t *ctx,
+                                unsigned char *output )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+    int ret = 0;
+    size_t olen = SASI_AES_BLOCK_SIZE_IN_BYTES;
+
+    if( ctx == NULL || ctx->cipher_info == NULL ||
+        ctx->cmac_ctx == NULL || output == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    if( cmac_ctx->is_cc_initiated == 0 )
+    {
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    if( ( ret = SaSi_AesFinish( &cmac_ctx->CC_Context, cmac_ctx->unprocessed_len,
+                                cmac_ctx->unprocessed_block,
+                                cmac_ctx->unprocessed_len, output, &olen ) ) != 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+        goto exit;
+    }
+
+exit:
+    if( SaSi_AesFree( &cmac_ctx->CC_Context ) != 0 && ret == 0 )
+    {
+        ret = MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED;
+    }
+
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
+
+    if( ctx == NULL || ctx->cipher_info == NULL || ctx->cmac_ctx == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    cmac_ctx = ctx->cmac_ctx;
+
+    /* Reset the internal state */
+    cmac_ctx->unprocessed_len = 0;
+    mbedtls_platform_zeroize( cmac_ctx->unprocessed_block,
+                              sizeof( cmac_ctx->unprocessed_block ) );
+
```

Comment:
Why doesnt this function need to reset the accelerator and clear the `initialized` flag.

Reply:
[because](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L134):
~~~
 * \brief               This function prepares the authentication of another
 *                      message with the same key as the previous CMAC
 *                      operation.
~~~
Since same key is used, no need to reset the accelerator

Reply:
ok